### PR TITLE
Added description of failure scenarios when code uses databases and brokers together

### PR DESCRIPTION
**Pull request by udidahan**
When developers are building mission-critical systems, they need to be aware of the various failure scenarios and how to handle them.

This PR gives them one detailed example and points them at resources that explain the topic further.

### CODE REVIEW

**Comment by udidahan** [comment]
@EdPrice-MSFT is this more the kind of thing you had in mind? Thanks!

**Comment by Jak-MS** [comment]
@msimecek 
- Can you review this PR?

IMPORTANT: When this content is ready to merge, you must add `#sign-off` in a comment or the approval may get overlooked.

cc: @EdPrice-MSFT 

#label:"aq-pr-triaged"
@MicrosoftDocs/public-repo-pr-review-team

File: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md [hunk 1]
```diff
@@ -157,6 +157,11 @@ Azure Service Bus premium tier is the recommended solution for high-value messag
...
+    - If the code emitting the messages runs before the database transaction is committed, which is how many developers work using the Unit of Work pattern, those messages can "escape" if the failure occurs between calling the broker and asking that the database transaction be committed. As the transaction rolls back, those database-generated IDs are also undone, leaving them available to other code which may be running at the same time. This can cause recipients of the "escaped" messages to processes the wrong database entries, hurting the overall consistency and correctness of your system.
+    - For a deeper discussion on these failure scenarios and how to handle them, see [this webinar on the topic](https://particular.net/webinars/implementing-an-outbox).
```
Comment (msimecek) [inline docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md:164]:
```suggestion
```

Comment (msimecek) [reply docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md:164]: Please remove this. A link to your company's webinar, promoting your product, is not part of Azure Mission-critical guidance.

File: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md [hunk 2]
```diff
@@ -157,6 +157,11 @@ Azure Service Bus premium tier is the recommended solution for high-value messag
...
+  - However, if the processing of the message includes database operations, specifically the insertion of new records with database-generated identifiers, as well as the emitting of new messages to the broker containing those identifiers, because there aren't distributed transactions which encompass both the database and the message broker, there can be a number of complications that can occur if the process running the code happens to fail, for example:
+    - If the code emitting the messages runs before the database transaction is committed, which is how many developers work using the Unit of Work pattern, those messages can "escape" if the failure occurs between calling the broker and asking that the database transaction be committed. As the transaction rolls back, those database-generated IDs are also undone, leaving them available to other code which may be running at the same time. This can cause recipients of the "escaped" messages to processes the wrong database entries, hurting the overall consistency and correctness of your system.
```
Comment (msimecek) [inline docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md:163]: This describes the complications, but does't provide guidance. What's the recommendation here?

**Review by msimecek — CHANGES_REQUESTED** [review]
Thank you for the contribution! A couple of comments from me.

**Comment by EdPrice-MSFT** [comment]
> @EdPrice-MSFT is this more the kind of thing you had in mind? Thanks!

I was thinking more like having you create a new architecture, like a Solution Idea, which shows how Azure services work with your company's solutions. 

Regarding this PR...

See feedback from @msimecek. Thanks, Martin!

**Comment by ckittel** [comment]
@udidahan do you plan on continuing this PR and following up with the suggestions from the current doc co-owner, @msimecek?

File: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md (hunk 2 above)
Comment (udidahan) [reply docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md:163]: Added recommendations.

File: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md (hunk 1 above)
Comment (udidahan) [reply docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md]: Done.

**Comment by udidahan** [comment]
I hope these changes address all the concerns raised. Thanks.

**Comment by ckittel** [comment]
@msimecek - please review when you have time.  Thanks!

File: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md [hunk 3]
```diff
@@ -157,6 +157,14 @@ Azure Service Bus premium tier is the recommended solution for high-value messag
...
+  - The solution to the above problems involves using the [Outbox pattern](/azure/architecture/best-practices/transactional-outbox-cosmos) where the outgoing messages are stored "off to the side" in the same transactional store as the business data and transmitted from there to the message broker when the initial message has been successfully processed.
+  - Since many developers are unfamiliar with these problems or their solutions, in order to guarantee that these techniques are applied consistenty in a mission-critical system, it is advisable to have the functionality of the outbox and the interaction with the message broker wrapped in some kind of library, and that all code processing and sending transactionally-significant messages make use of that library rather than interacting with the message broker directly.
+    - Libraries that implement this functionality in .net include [NServiceBus](https://docs.particular.net/nservicebus/outbox/) and [MassTransit](https://masstransit-project.com/advanced/transactional-outbox.html).
```
Comment (msimecek) [inline docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md:167]:
```suggestion
    - Libraries that implement this functionality in .NET include [NServiceBus](https://docs.particular.net/nservicebus/outbox/) and [MassTransit](https://masstransit-project.com/advanced/transactional-outbox.html).
```

File: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md [hunk 4]
```diff
@@ -157,6 +157,14 @@ Azure Service Bus premium tier is the recommended solution for high-value messag
...
+    - If developers put the code which emits the message *after* the database transaction completes, if the process fails right between the two of them, when the message gets processed again, the idempotence guard clause which checks whether the message was processed before will see that it was (based on the data stored in the database) and then skip the message emitting code, believing that everything was done successfully last time. In this manner, downstream systems which were expecting to receive notifications about the completed process do not receive anything, once again resulting in an overall state of inconsistency.
```
Comment (msimecek) [inline docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md:164]: This is important, but very complicated sentence. I tried to reword it a little but, but also preserve the information:

```suggestion
    - If developers put the code which emits the message *after* the database transaction completes, the process can still fail between these operations (transaction commited - message sent). When that happens, the message will go through processing again, but this time the idempotence guard clause will see that it has been already processed (based on the data stored in the database) and will skip the message emitting code, believing that everything was done successfully last time. Downstream systems which were expecting to receive notifications about the completed process do not receive anything, once again resulting in an overall state of inconsistency.
```

**Review by msimecek — CHANGES_REQUESTED** [review]
Thanks for the updates! Two more adjustmens from me.

File: docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md [hunk 5]
```diff
@@ -157,6 +157,14 @@ Azure Service Bus premium tier is the recommended solution for high-value messag
   - If an acknowledgment isn't received by the broker in the allotted time period, or the handler explicitly abandons the message, the exclusive lock is released. The message is then available for other consumers to process the message.
   - If a message is not successfully processed a configurable number of times, or the handler forwards the message to the dead-letter queue.
 - Because messages can potentially be processed more than one time, message handlers should be made idempotent.
+  - "A ... method is considered `idempotent` if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request." ([RFC 7231](https://tools.ietf.org/html/rfc7231#section-4))
```
Comment (ckittel) [inline docs/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform.md:160]:
```suggestion
  - "A ... method is considered _idempotent_ if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request." ([RFC 7231](https://tools.ietf.org/html/rfc7231#section-4))
```

**Comment by ckittel** [comment]
Added a commit just to see if I can get the build to kick off again (might have an upstream problem here)

**Comment by ckittel** [comment]
> Some unexpected errors happened when running build, please open a ticket in https://aka.ms/SiteHelp and include the error report for our team to troubleshoot

@udidahan -- looks like our build process is currently being temperamental.  Nothing for you to address here, we'll keep trying.

**Comment by udidahan** [comment]
Thanks for letting me know. I was a bit worried 😅

**Review by EdPrice-MSFT — APPROVED** [review]
Finished edit. Looks great!

**Comment by EdPrice-MSFT** [comment]
I created an incident to resolve the build error: https://portal.microsofticm.com/imp/v3/incidents/details/336037212/home

**Review by EdPrice-MSFT — APPROVED** [review]
Made an edit pass. Please review.

Looks great!

**Comment by EdPrice-MSFT** [comment]
#Sign-Off

**Comment by udidahan** [comment]
Thanks everyone for helping get this across the finish line.

BTW, how long does it usually take for these merges to make their way to the public site? The reason I ask is that I don't see the changes on the public site:

https://learn.microsoft.com/en-us/azure/architecture/reference-architectures/containers/aks-mission-critical/mission-critical-data-platform#every-message-must-be-processed

**Comment by udidahan** [comment]
OK. I see that it's live now. Thank you all.

**Pull request merged**